### DISCUSSION TITLE
Failed to load module "colorreload-gtk-module"

### DISCUSSION THREAD

**Discussion by C0rn3j** [pyinstaller/pyinstaller#8945] [Help] [answered]
On Arch Linux in a Plasma Wayland session.

There is this file `/usr/lib/gtk-3.0/modules/libcolorreload-gtk-module.so` - there's also `libwindow-decorations-gtk-module.so` I have an identical issue with.

```
Gtk-Message: 02:19:03.187: Failed to load module "colorreload-gtk-module"
Gtk-Message: 02:19:03.187: Failed to load module "window-decorations-gtk-module"
```

I build on Ubuntu 24.04, and whenever I try to launch on Arch Linux, I get these (seemingly harmless) warnings.

I tried harcoding them in, even putting them in gtk-3.0 subfolder but no matter what I do I get these module warnings.

They're apparently defined in `~/.config/gtk-3.0/settings.ini` but not sure if that's any helpful.

```python
	datas=[
		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libcolorreload-gtk-module.so", "."),
		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libwindow-decorations-gtk-module.so", "."),
```

How could I get rid of these warnings?

I wanted to try this locally, but Arch just updated to Python 3.13, I am no longer able to run it successfully - build log from that is here https://paste.rys.rs/?62b0176e39236009#Dfoj4DstzXBkNHgF1Xzu5QG67BuB3vkphEZ4mcCXwkyp

But the gist from that one seems to be:

```python
  File "/usr/lib/python3.13/site-packages/PyInstaller/hooks/hook-setuptools.py", line 52, in <module>
    if setuptools_info.version < (71, 0):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '<' not supported between instances of 'NoneType' and 'tuple'
```

Logs + build from the problematic app can be found here - https://github.com/Taiko2k/Tauon/actions/runs/12459392309

**Comment by rokm** [comment]
> On Arch Linux in a Plasma Wayland session.
> 
> There is this file `/usr/lib/gtk-3.0/modules/libcolorreload-gtk-module.so` - there's also `libwindow-decorations-gtk-module.so` I have an identical issue with.
> 
> ```
> Gtk-Message: 02:19:03.187: Failed to load module "colorreload-gtk-module"
> Gtk-Message: 02:19:03.187: Failed to load module "window-decorations-gtk-module"
> ```
> 
> I build on Ubuntu 24.04, and whenever I try to launch on Arch Linux, I get these (seemingly harmless) warnings.
> 
> I tried harcoding them in, even putting them in gtk-3.0 subfolder but no matter what I do I get these module warnings.
> 
> They're apparently defined in `~/.config/gtk-3.0/settings.ini` but not sure if that's any helpful.
> 
> ```python
> 	datas=[
> 		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libcolorreload-gtk-module.so", "."),
> 		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libwindow-decorations-gtk-module.so", "."),
> ```
> 
> How could I get rid of these warnings?

I imagine these modules need to be collected into `gtk-3.0/modules` sub-directory instead of top-level application directory. Depending on how their search is implemented in the Gtk, you might also need to explicitly override the search path via some environment variable.

**Reply by C0rn3j** [reply]
I have tried basically all combinations of the directories, it probably needs an override I so far haven't been able to find, will try to look harder.

**Reply by rokm** [reply]
I suspect you cannot fully fix these sort of warnings anyway, because the user's system might be using Gtk modules that your application does not provide. (So the warnings are result of mismatch between system's Gtk installation and its configuration vs Gtk bundled with the application, which is also using system's Gtk configuration).

**Reply by C0rn3j** [reply]
https://docs.gtk.org/gtk3/running.html

`GTK_DEBUG="modules" GTK3_MODULES="libcolorreload-gtk-module" ./Tauon\ Music\ Box` I have at least managed to force the error this way, but DEBUG does not work because Ubuntu does not build GTK with DEBUG it seems, will try further when it's possible to run on Arch Linux again and I can use the debug logs.

**Reply by rokm** [reply]
FWIW, our Gtk run-time hook overrides `GTK_PATH` to the application's top-level application directory:
https://github.com/pyinstaller/pyinstaller/blob/c2dba7f115705f150b3eb1f5e7ddcf6c669873bf/PyInstaller/hooks/rthooks/pyi_rth_gtk.py#L17-L23

I guess the question is whether Gtk expects the modules to be in a sub-directory (` gtk-3.0/modules`, or perhaps even just `modules`) or not...

**Reply by C0rn3j** [reply, marked as answer]
```
GTK_PATH
Specifies a list of directories to search when GTK is looking for dynamically loaded objects such as the modules specified by GTK_MODULES, theme engines, input method modules, file system backends and print backends. 

If the path to the dynamically loaded object is given as an absolute path name, then GTK loads it directly. 

Otherwise, GTK goes in turn through the directories in GTK_PATH, followed by the directory .gtk-3.0 in the user’s home directory, followed by the system default directory, which is libdir/gtk-3.0/modules. 

(If GTK_EXE_PREFIX is defined, libdir is $GTK_EXE_PREFIX/lib. Otherwise it is the libdir specified when GTK was configured, usually /usr/lib, or /usr/local/lib.) 

For each directory in this list, GTK actually looks in a subdirectory directory/version/host/type where version is derived from the version of GTK (use pkg-config --variable=gtk_binary_version gtk+-3.0 to determine this from a script); host is the architecture on which GTK was built. 

(use pkg-config --variable=gtk_host gtk+-3.0 to determine this from a script); and type is a directory specific to the type of modules; currently it can be:
```

The important part:
`Otherwise, GTK goes in turn through the directories in GTK_PATH, followed by the directory .gtk-3.0 in the user’s home directory, followed by the system default directory, which is libdir/gtk-3.0/modules. (If GTK_EXE_PREFIX is defined, libdir is $GTK_EXE_PREFIX/lib`

I missed the `lib` prefix, it seems to register now!

```
	datas=[
		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libcolorreload-gtk-module.so", "lib/gtk-3.0/modules"),
		("/usr/lib/x86_64-linux-gnu/gtk-3.0/modules/libwindow-decorations-gtk-module.so", "lib/gtk-3.0/modules"),
```

**Comment by rokm** [comment]
> I wanted to try this locally, but Arch just updated to Python 3.13, I am no longer able to run it successfully - build log from that is here https://paste.rys.rs/?62b0176e39236009#Dfoj4DstzXBkNHgF1Xzu5QG67BuB3vkphEZ4mcCXwkyp
> 
> But the gist from that one seems to be:
> 
> ```python
>   File "/usr/lib/python3.13/site-packages/PyInstaller/hooks/hook-setuptools.py", line 52, in <module>
>     if setuptools_info.version < (71, 0):
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: '<' not supported between instances of 'NoneType' and 'tuple'
> ```

This second part is a bug. It looks like Arch-packaged `python-setuptools` removes the `setuptools._vendor` (since it installs those dependencies as external packages), and we do not account for that possibility. With `setuptools`-vendored dependencies becoming optional, it kind of makes sense for linux distributions to remove the `_vendor` directory.

There's an error earlier in your build log,
>``` 
>6465 WARNING: SetuptoolsInfo: failed to obtain setuptools info: Child process call to _retrieve_setuptools_info() failed with:
>  File "/usr/lib/python3.13/site-packages/PyInstaller/isolated/_child.py", line 63, in run_next_command
>    output = function(*args, **kwargs)
>  File "/usr/lib/python3.13/site-packages/PyInstaller/utils/hooks/setuptools.py", line 82, in _retrieve_setuptools_info
>    setuptools_vendor = importlib.import_module("setuptools._vendor")
>  File "/usr/lib/python3.13/importlib/__init__.py", line 88, in import_module
>    return _bootstrap._gcd_import(name[level:], package, level)
>           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
>  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
>  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
>ModuleNotFoundError: No module named 'setuptools._vendor'
>``` 

which probably prevents the `setuptools_info` from being completely initialized, leading to subsequent errors.

Will fix this.

**Reply by rokm** [reply]
#8947 should take care of this issue.
